File: test_spacesaving.py
import string
import random
import time

import numpy as np

import bloom_filter as bf

def rand_id_list(len=1000):
    return np.random.randint(0, 1024, len);

def skewed_freq(len=1000):
    return 20 * np.random.zipf(2, len);

if __name__ == '__main__':
    sketch = bf.spacesaving(31);

    rand_array = rand_id_list(10240);
    rand_freq = skewed_freq(10240);

    for i in range(0, 10240):
        sketch.insert(rand_array[i], rand_freq[i]);

    res = sketch.query_top();

File: test_spacesaving.cpp
#include "spacesaving.h"

#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int main(){
    space_saving ss(500);
    vector<int> data(10240);
    for(int i = 0; i < 10240; i += 1){
    	data[i] = i + 1;
	}
	random_shuffle(data.begin(), data.end());
	for(int i = 0; i < 10240; i += 1){
		ss.insert(data[i], 10240 / data[i]);
	}
    vector<node> res(500);
    ss.query_top(res);
    // works better on skewed data
    for(int i = 0; i < 500; i += 1){
    	cout << res[i].value << endl;
	}
    return 0;
}
